File: code/Java/src/main/java/App/CommandInterpreter.java
package App;
import Commands.*;
import Model.Coordinate;
import Model.Direction;
import Model.Position;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import static App.Configuration.Primitives.*;
import static java.lang.Integer.parseInt;

public class CommandInterpreter {

    private Map<String, Direction> letterToDirection = new HashMap<>() {{
        put(NORTH, Direction.NORTH);
        put(EAST, Direction.EAST);
        put(SOUTH, Direction.SOUTH);
        put(WEST, Direction.WEST);
    }};

    public ArrayList<ICommand> translate(String commands) {

        ArrayList<ICommand> allCommands = new ArrayList<>();
        allCommands.add(getInitialisationCommand(commands));
        allCommands.add(getStartingPositionCommand(commands));
        allCommands.addAll(getMovementCommands(commands));

        return allCommands;
    }

    private ArrayList<ICommand> getMovementCommands(String commands) {
        ArrayList<ICommand> movementCommands = new ArrayList<>();
        String[] lines = commands.split("\n");

        for (Character command : lines[2].toCharArray()){
            switch (command){
                case TURN_LEFT -> movementCommands.add(new TurnLeftCommand());
                case MOVE_FORWARD -> movementCommands.add(new MoveForwardCommand());
                case TURN_RIGHT -> movementCommands.add(new TurnRightCommand());
            }
        }
        return movementCommands;
    }

    private InitialisationCommand getInitialisationCommand(String commands) {
        String[] lines = commands.split("\n");
        String[] topRight = lines[0].split(" ");

        return new InitialisationCommand(new Coordinate(parseInt(topRight[0]), parseInt(topRight[1])));
    }

    private StartingPositionCommand getStartingPositionCommand(String commands) {
        String[] lines = commands.split("\n");
        String[] coords = lines[1].split(" ");

        Coordinate coordinate = new Coordinate(parseInt(coords[0]), parseInt(coords[1]));
        Direction direction = letterToDirection.get(coords[2]);
        Position position = new Position(coordinate, direction);

        return new StartingPositionCommand(position);
    }
}

File: code/CSharp/ConnascentMarsRover/Source/App/CommandInterpreter.cs
using System.Collections.Generic;
using Source.Commands;
using Source.Model;
using static Source.App.Configuration.Primitives;

namespace Source.App
{
    public class CommandInterpreter
    {
        private const string NORTH = "N";
        private const string EAST = "E";
        private const string SOUTH = "S";
        private const string WEST = "W";

        private readonly Dictionary<string, Direction> _letterToDirection =
            new Dictionary<string, Direction>
            {
                {NORTH, Direction.NORTH},
                {EAST, Direction.EAST},
                {SOUTH, Direction.SOUTH},
                {WEST, Direction.WEST}
            };


        public List<ICommand> Translate(string commands)
        {
            var allCommands = new List<ICommand>();
            allCommands.Add(GetInitialisationCommand(commands));
            allCommands.Add(GetStartingPositionCommand(commands));
            allCommands.AddRange(GetMovementCommands(commands));

            return allCommands;
        }

        private IEnumerable<ICommand> GetMovementCommands(string commands)
        {
            var movementCommands = new List<ICommand>();
            var lines = commands.Split("\n");

            foreach (var command in lines[2])
            {
                switch (command)
                {
                    case TURN_LEFT:
                        movementCommands.Add(new TurnLeftCommand());
                        break;
                    case MOVE_FORWARD:
                        movementCommands.Add(new MoveForwardCommand());
                        break;
                    case TURN_RIGHT:
                        movementCommands.Add(new TurnRightCommand());
                        break;
                }
            }

            return movementCommands;
        }

        private InitialisationCommand GetInitialisationCommand(string commands)
        {
            var lines = commands.Split("\n");
            var topRight = lines[0].Split(" ");

            return new InitialisationCommand(
                new Coordinate(int.Parse(topRight[0]), int.Parse(topRight[1])));
        }

        private StartingPositionCommand GetStartingPositionCommand(string commands)
        {
            var lines = commands.Split("\n");
            var coords = lines[1].Split(" ");

            var coordinate = new Coordinate(int.Parse(coords[0]), int.Parse(coords[1]));
            var direction = _letterToDirection[coords[2]];
            var position = new Position(coordinate, direction);

            return new StartingPositionCommand(position);
        }
    }
}